File: soomin/후보키.cpp
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

bool check_minimality(const vector<int> &keys, const int bit)
{
    for (int i = 0; i < keys.size(); ++i)
    {
        if ((keys[i] & bit) == keys[i]) //비트의 속성 이용. 같다면 새로운 키가 기존 키를 포함. 최소성 불만족.
        {
            return false;
        }
    }
    return true;
}

int solution(vector<vector<string>> relation) {

    vector<int> keys;
    int cnt = 0;
    int row = relation.size();
    int col = relation[0].size();

    // 후보키 속성 구성을 비트마스크 순열로 조합하기
    for (int bit = 1; bit < (1 << col); ++bit)
    {
        if (!check_minimality(keys, bit))
        {
            continue;
        }

        // 유일성
        unordered_set<string> set;
        for (int i = 0; i < row; ++i)
        {
            string key;
            for (int j = 0; j < col; ++j)
            {
                if (bit & (1 << j))
                {
                    key += relation[i][j] + ' ';    // 속성간 구분자 추가
                }
            }
            set.insert(key);
        }

        if (set.size() == row) //겹치는 것이 없고 하나씩 모두 있다면 유일성 만족
        {
            keys.push_back(bit);
            ++cnt;
        }
    }
    
    return cnt;
}

File: minnkyoung/후보키.py
from itertools import combinations

def solution(relation):
    
    answer = 0
    
    all = list() #모든 col의 조합 리스트
    uniqeIndex = [] #유일성을 만족하는 조합 리스트
    
    if len(relation) > 0:

        colSize = len(relation[0]) #col개수

        rowSize = len(relation) #row개수

        for i in range(1, colSize + 1): #모든 col의 조합 구하기(set)

            all.extend([set(k) for k in combinations([j for j in range(colSize)], i)])

        for comb in all: #조합들의 유일성 검사

            vaildSet = set() #set에 추가하여 원래의 row크기와 비교

            for row in range(rowSize): #조합에 해당되는 row를 하나의 문자열로 만들어 set에 넣기
                temp = ''
    
                for col in comb:
                    temp += relation[row][col]
                
                vaildSet.add(temp)

            if len(vaildSet) == rowSize: 
                uniqeIndex.append(comb)

    delSet = set() #최소성을 위배하는 set

    for i in uniqeIndex:
        for idx, j in enumerate(uniqeIndex):

            if i.issubset(j) and i != j:
                delSet.add(uniqeIndex.index(j))

    answer = len(uniqeIndex)-len(delSet)
    
    return answer